shell/browser/api/atom_api_web_request_ns.cc

```diff
@@ -94,15 +191,62 @@ void WebRequestNS::SetListener(Event event, gin::Arguments* args) {
     return;
   }
 
-  // TODO(zcbenz): Actually set the listeners.
-  args->ThrowTypeError("This API is not implemented yet");
+  if (listener.is_null())
+    listeners->erase(event);
+  else
+    (*listeners)[event] = {std::move(patterns), std::move(listener)};
+}
+
+template <typename... Args>
+void WebRequestNS::HandleSimpleEvent(SimpleEvent event,
+                                     extensions::WebRequestInfo* request,
+                                     Args... args) {
+  const auto& info = simple_listeners_[event];
+  if (!MatchesFilterCondition(request, info.url_patterns))
+    return;
+
+  // TODO(zcbenz): Invoke the listener.
+}
+
+template <typename Out, typename... Args>
+int WebRequestNS::HandleResponseEvent(ResponseEvent event,
+                                      extensions::WebRequestInfo* request,
+                                      net::CompletionOnceCallback callback,
+                                      Out out,
+                                      Args... args) {
+  const auto& info = response_listeners_[event];
+  if (!MatchesFilterCondition(request, info.url_patterns))
+    return net::OK;
+
+  // TODO(zcbenz): Invoke the listener.
+  return net::OK;
 }
 
 // static
-gin::Handle<WebRequestNS> WebRequestNS::Create(
+gin::Handle<WebRequestNS> WebRequestNS::FromOrCreate(
     v8::Isolate* isolate,
-    AtomBrowserContext* browser_context) {
-  return gin::CreateHandle(isolate, new WebRequestNS(isolate, browser_context));
+    content::BrowserContext* browser_context) {
+  auto* web_request = From(browser_context);
+  if (!web_request) {
+    // Create.
+    web_request = new WebRequestNS(isolate, browser_context);
+    // Make sure the |Session| object has the |webRequest| property managed.
+    Session::CreateFrom(isolate,
```

Comment:
This seems circular, `Session::WebRequest` => `WebRequest::FromOrCreate` => `Session::WebRequest`.